src/main/scala/ai/privado/languageEngine/csharp/tagger/collection/CollectionTagger.scala

```diff
@@ -0,0 +1,139 @@
+
+import scala.collection.mutable
+
+class CollectionTagger(cpg: Cpg, ruleCache: RuleCache) extends PrivadoParallelCpgPass[RuleInfo](cpg) {
+  protected val methodUrlMap: mutable.HashMap[Long, String] = mutable.HashMap[Long, String]()
+  protected val classUrlMap: mutable.HashMap[Long, String]  = mutable.HashMap[Long, String]()
+  private val logger                                        = LoggerFactory.getLogger(this.getClass)
+
+  override def generateParts(): Array[RuleInfo] =
+    ruleCache.getRule.collections.filter(_.catLevelTwo == Constants.annotations).toArray
+
+  override def runOnPart(builder: DiffGraphBuilder, ruleInfo: RuleInfo): Unit = {
+    classUrlMap.addAll(collectAnnotatedUrlsFromClasses(ruleInfo.combinedRulePattern))
+    val (methodUrls, collectionMethodsCache) = collectAnnotatedUrlsFromMethods(ruleInfo.combinedRulePattern)
+    methodUrlMap.addAll(methodUrls)
+    tagSources(builder, ruleInfo, collectionMethodsCache)
+  }
+
+  protected def tagSources(
+    builder: DiffGraphBuilder,
+    ruleInfo: RuleInfo,
+    collectionMethodsCache: List[Method]
+  ): Unit = {
+    CollectionUtility.tagDirectSources(
+      builder,
+      collectionMethodsCache,
+      ruleCache.getRule.sources,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+
+    CollectionUtility.tagDerivedSources(
+      cpg,
+      builder,
+      collectionMethodsCache,
+      ruleInfo,
+      ruleCache,
+      methodUrlMap = methodUrlMap,
+      classUrlMap = classUrlMap
+    )
+  }
+
+  private def getUrlFromAnnotationsCode(annotation: Annotation): String = {
+    // If we have [controller] in annotation code, then replace with the class name of controller
+    //
+    //    [Route("api/some/[controller]")]                      <-- /api/some/email
+    //    class EmailController {
+    //      [HttpGet("[controller]/[action]/{id}")]             <-- /api/some/email/edit/{id}
+    //      public IActionResult Edit(int id)
+    //      {
+    //        return ControllerContext.MyDisplayRouteInfo(id);
+    //      }
+    //    }
+    //
+    val controllerName = if annotation.code.contains("[controller]") then {
+      cpg.typeDecl
+        .where(_.annotation.codeExact(annotation.code))
+        .where(_.name(".*Controller"))
+        .name
+        .headOption
+        .getOrElse("")
+        .split("Controller")
+        .headOption
+        .getOrElse("")
+        .toLowerCase
+    } else {
+      ""
+    }
+
+    // Now we extract [action] seen in snippet above. Only method support for now since we encounter them most here
+    // TODO: Support [action] replacement for classes as well eg Route("[controller]/[action]")
+    val actionName = if annotation.code.contains("[action]") then {
+      cpg.method.where(_.annotation.codeExact(annotation.code)).name.headOption.getOrElse("").toLowerCase
+    } else {
+      ""
+    }
+    annotation.code
+      .replaceAll(".*\"(.*?)\".*", "/$1")
```

Comment:
Here are we looking to pick the content present inside the `""` double quotes?
Please add a comment explaining the purpose of this